### ISSUE TITLE
add support for tuning hyper-parameter sigma_s when using semi-parametric (2D-AR) selectivity

### ISSUE THREAD

**Issue by iantaylor-NOAA** [r4ss/r4ss#500]
Except below from [Xu et al. (2018)](https://cdnsciencepub.com/doi/10.1139/cjfas-2017-0446) describes a tuning algorithm for the hyper-parameter sigma_s used with semi-parametric selectivity (which has a label like "sigmasel_Fishery" in SS).

A table similar to `$sigma_R_info` (discussed in issue #120) could be produced for these models to help with tuning. We could also add a function or expand `SS_tune_comps()` to automatically iterate the value based on this approach.

@HaikunXu and @James-Thorson, let me know if you have any thoughts on the best way to do this.

![image](https://user-images.githubusercontent.com/4992918/112058559-f83a2180-8b17-11eb-818e-cf4a3289e28f.png)

**Comment by James-Thorson-NOAA** [comment]
Do you know if anyone is using the semi-parametric selectivity features ... is it "operational" for hake or some other species?  

I would think it's worth extra time for this r4ss infrastructure if it's being used operationally ...?

**Comment by iantaylor-NOAA** [comment]
Good question @James-Thorson.
This issue was inspired by an Andre Punt email where he wrote "I am wondering - you don't compute the 'adjust' sigma for the 2D selectivity mode (aka Xu et al?)" and later "Yes, re-running would be ideal" (meaning automated iterative tuning within r4ss like that added for sample sizes by @k-doering-NOAA in pull request https://github.com/r4ss/r4ss/pull/439).

The context suggests that Andre's application is a simulation analysis, not a production stock assessment, but I think that's the path that these things SHOULD take first.

Combining this with issue https://github.com/r4ss/r4ss/issues/498 makes me think that the ideal would be a single function that can tune all elements of a model that require iteration. It would take some work to get there, but it seems useful to have that vision in mind while we work toward the day where we can fully simplify stock assessment by replacing tuning algorithms with statistics.

**Comment by James-Thorson-NOAA** [comment]
well, I certainly support the idea of doing a simulation test of these estimation approaches for hyperparameters!  

I say go for it :)  My only other note is that @kellijohnson-NOAA might be automating the previous (little-used) code for the Laplace approximation, as in the Xu et al. paper that did well too, and would generalize more easily to jointly tuning selectivity and other hyperparameters.

**Comment by HaikunXu** [comment]
Yes, I have sent @kellijohnson-NOAA the cpp code for estimating the two autocorrelation parameters for selectivity deviations. It would make the usage of semi-parametric selectivity much more convenient if r4ss can incorporate that code to report estimated autocorrelation parameters as well (in addition to sigmasel_Fishery).

**Comment by iantaylor-NOAA** [comment]
Here is code kindly provided by Andre Punt which we can work on incorporating in the future:
```
TuneSelSigma <- function(mymodel)
{
  # Implemments equations 19 and 20 Xu et al. (2019)
 
  # Number of fleets and their names
  Nfleet <- length(mymodel$definitions[,1])  
  Fleets <- mymodel$definitions$Fleet_name
 
  sdSel <- rep(NA,Nfleet); b <- rep(NA,Nfleet)
  for (Ifleet in 1:Nfleet)
  {
    FleetLen <- nchar(Fleets[Ifleet])
    ParNames <- substr(mymodel$parameters[,2],1,FleetLen+6)
    Use <- which(ParNames==paste(Fleets[Ifleet],"_ARDEV",sep=""))
    if (length(Use)>0)
    {  
      VarEps <- var(mymodel$parameters$Value[Use])
      Sds <- mymodel$parameters$Parm_StDev[Use]
      VarEpsEst <- sum(Sds^2)/length(Sds)
      sdSel[Ifleet] <- sqrt(VarEps+VarEpsEst)
      b[Ifleet] <- 1.0 - VarEpsEst/sdSel[Ifleet]^2
      cat(Fleets[Ifleet],VarEps,VarEpsEst,sdSel[Ifleet],b[Ifleet],"\n")
    }
  }
  Outs <- NULL
  Outs$sdSel <- sdSel
  Outs$b <- b
  return(Outs)
}
```
